sahi/utils/cv.py

```diff
@@ -210,6 +210,78 @@ def apply_color_mask(image: np.ndarray, color: tuple):
     return colored_mask
 
 
+def get_video_reader(
+    source: str, save_dir: str, export_visual: bool = False, view_image: bool = False, fast_forwarding: bool = False
+):
+    # get video name with extension
+    image_base = os.path.basename(source)
+    # get video from video path
+    video_capture = cv2.VideoCapture(source)
+    if export_visual:
+        # get video properties and create VideoWriter object
+        fps = video_capture.get(cv2.CAP_PROP_FPS)
+        w = int(video_capture.get(cv2.CAP_PROP_FRAME_WIDTH))
+        h = int(video_capture.get(cv2.CAP_PROP_FRAME_HEIGHT))
+        size = (w, h)
+        fourcc = cv2.VideoWriter_fourcc(*"mp4v")
+        out = cv2.VideoWriter(os.path.join(save_dir, image_base), fourcc, fps, size)
+
+        if view_image:
+            cv2.imshow("Prediction of {}".format(str(image_base)), cv2.WINDOW_AUTOSIZE)
+
+    if fast_forwarding:
+
+        def read_video_frame(video_capture):
+            while video_capture.isOpened:
+                k = cv2.waitKey(30)
+
+                frame_num = video_capture.get(cv2.CAP_PROP_POS_FRAMES)
+
+                if k == 27:
+                    print(
+                        "\n===========================Closing==========================="
+                    )  # Exit the prediction, Key = Esc
+                    exit()
+                if k == 100:
+                    frame_num += 100  # Skip 100 frames, Key = d
+                if k == 97:
+                    frame_num -= 100  # Prev 100 frames, Key = a
+                if k == 103:
+                    frame_num += 10  # Skip 10 frames, Key = g
+                if k == 102:
+                    frame_num -= 10  # Prev 10 frames, Key = f
+                video_capture.set(cv2.CAP_PROP_POS_FRAMES, frame_num)
+
+                ret, frame = video_capture.read()
+                if not ret:
+                    print("\n===========================Video Ended===========================")
+                    break
+                yield Image.fromarray(frame)
+
+    else:
+
+        def read_video_frame(video_capture):
+            while video_capture.isOpened:
+                k = cv2.waitKey(30)
+
+                if k == 27:
+                    print(
+                        "\n===========================Closing==========================="
+                    )  # Exit the prediction, Key = Esc
+                    exit()
+
+                ret, frame = video_capture.read()
+                if not ret:
+                    print("\n===========================Video Ended===========================")
+                    break
+                yield Image.fromarray(frame)
+
+    if export_visual:
```

Comment:
Let's delete the if blocks here and only return the following (see also the comment above for `out`) that way we unify the return type which makes the code less prone to mistakes.

```python
return read_video_frame(video_capture), out, image_base
```

Reply:
If we don't use two option of return, every run sahi, it was creating null output file in `runs/predict`. Because of `out` variable.

Reply:
I changed it but I would have to use if block for return.

Reply:
why tho ? Isn't this controlled with `export_visual`, so my thought was like this

```python
if export_visual:
    # Regular stuff here
    # get video properties and create VideoWriter object
    fps = video_capture.get(cv2.CAP_PROP_FPS)
    # ...
else:
    out = None

return read_video_frame(video_capture), out, image_base
```

If you think this would behave to create empty output file, can you elaborate why ?

Reply:
I am interested in this situation.